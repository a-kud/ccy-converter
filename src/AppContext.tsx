import { createContext, useReducer } from "react";
import { getNextIndex, getPreviousIndex } from "./utils/list";
import {
  addAmount,
  convertAmount,
  invertRate,
  parseAmountInput,
  subtractAmount,
} from "./utils/number";

export interface AppState {
  accounts: AccountData[];
  baseAccountIdx: number;
  quoteAccountIdx: number;
  baseAmount: string;
  quoteAmount: string;
  baseActive: boolean;
  rate: string;
}

export interface AccountData {
  currency: string;
  amount: string;
}

export function initialState(): AppState {
  const accounts = [
    { currency: "EUR", amount: "9999" },
    { currency: "USD", amount: "9000" },
    { currency: "GBP", amount: "42" },
  ];

  return {
    accounts,
    baseAccountIdx: 0,
    quoteAccountIdx: 1,
    baseAmount: "",
    quoteAmount: "",
    baseActive: true,
    rate: "0",
  };
}

interface AppContextValue {
  state: AppState;
  dispatch: (action: Action) => void;
}

interface RateChangedAction {
  type: "RATE_CHANGED";
  payload: {
    rate: string;
  };
}

interface BaseAccountChangedAction {
  type: "BASE_CHANGED";
  payload: {
    next?: boolean;
  };
}

interface QuoteAccountChangedAction {
  type: "QUOTE_CHANGED";
  payload: {
    next?: boolean;
  };
}

interface BaseAmountChangedAction {
  type: "BASE_AMOUNT_CHANGED";
  payload: {
    amount: string;
  };
}

interface QuoteAmountChangedAction {
  type: "QUOTE_AMOUNT_CHANGED";
  payload: {
    amount: string;
  };
}

interface ExchangeAction {
  type: "EXCHANGE_REQUEST";
  payload: {};
}

type Action =
  | RateChangedAction
  | BaseAccountChangedAction
  | QuoteAccountChangedAction
  | BaseAmountChangedAction
  | QuoteAmountChangedAction
  | ExchangeAction;

export const AppContext = createContext<AppContextValue>({
  state: initialState(),
  dispatch: () => {},
});

export function AppContextProvider({
  children,
}: {
  children: React.ReactChild;
}) {
  const [state, dispatch] = useReducer(reducer, initialState());

  const contextValue: AppContextValue = {
    state,
    dispatch,
  };
  return (
    <AppContext.Provider value={contextValue}>{children}</AppContext.Provider>
  );
}

export function reducer(state: AppState, action: Action): AppState {
  switch (action.type) {
    case "RATE_CHANGED":
      return rateChangeReducer(state, action.payload);
    case "BASE_CHANGED":
      return baseChangeReducer(state, action.payload);
    case "QUOTE_CHANGED":
      return quoteChangeReducer(state, action.payload);
    case "BASE_AMOUNT_CHANGED":
      return baseAmountChangeReducer(state, action.payload);
    case "QUOTE_AMOUNT_CHANGED":
      return quoteAmountChangeReducer(state, action.payload);
    case "EXCHANGE_REQUEST":
      return exchangeReducer(state);
    default:
      return state;
  }
}

function exchangeReducer(state: AppState): AppState {
  const { accounts, baseAccountIdx, quoteAccountIdx, baseAmount, quoteAmount } =
    state;
  const baseAccount = accounts[baseAccountIdx];
  const quoteAccount = accounts[quoteAccountIdx];
  if (!baseAccount || !quoteAccount) {
    return state;
  }

  baseAccount.amount = subtractAmount(baseAccount.amount, baseAmount);
  quoteAccount.amount = addAmount(quoteAccount.amount, quoteAmount);

  const updatedAccounts = accounts.map((acc) => {
    if (acc.currency === baseAccount.currency) {
      return { ...baseAccount };
    }

    if (acc.currency === quoteAccount.currency) {
      return { ...quoteAccount };
    }

    return acc;
  });

  return {
    ...state,
    accounts: updatedAccounts,
    baseAmount: "",
    quoteAmount: "",
  };
}

function baseChangeReducer(
  state: AppState,
  payload: BaseAccountChangedAction["payload"]
): AppState {
  const { next } = payload;
  const { baseAccountIdx, accounts } = state;
  const newIdx = next
    ? getNextIndex(accounts, baseAccountIdx)
    : getPreviousIndex(accounts, baseAccountIdx);

  return { ...state, baseAccountIdx: newIdx, rate: "0" };
}

function quoteChangeReducer(
  state: AppState,
  payload: QuoteAccountChangedAction["payload"]
): AppState {
  const { next } = payload;
  const { quoteAccountIdx, accounts } = state;

  const newIdx = next
    ? getNextIndex(accounts, quoteAccountIdx)
    : getPreviousIndex(accounts, quoteAccountIdx);

  return { ...state, quoteAccountIdx: newIdx, rate: "0" };
}

function baseAmountChangeReducer(
  state: AppState,
  payload: BaseAmountChangedAction["payload"]
): AppState {
  const baseAmount = parseAmountInput(payload.amount);
  return {
    ...state,
    baseAmount: baseAmount,
    quoteAmount: convertAmount(baseAmount, state.rate),
    baseActive: true,
  };
}

function quoteAmountChangeReducer(
  state: AppState,
  payload: QuoteAmountChangedAction["payload"]
): AppState {
  const quoteAmount = parseAmountInput(payload.amount);
  return {
    ...state,
    quoteAmount: quoteAmount,
    baseAmount: convertAmount(quoteAmount, invertRate(state.rate)),
    baseActive: false,
  };
}

function rateChangeReducer(
  state: AppState,
  payload: RateChangedAction["payload"]
): AppState {
  const { rate } = payload;
  if (rate === state.rate) {
    return state;
  }

  let { baseAmount, quoteAmount, baseActive } = state;

  if (baseActive) {
    quoteAmount = convertAmount(baseAmount, rate);
  } else {
    baseAmount = convertAmount(quoteAmount, invertRate(rate));
  }

  return { ...state, rate: rate, baseAmount, quoteAmount };
}
