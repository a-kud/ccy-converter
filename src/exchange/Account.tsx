import React from "react";
import styled from "styled-components";
import { color } from "../styles";
import { ExchangeRateType } from "../types";
import { AmountInput } from "./AmountInput";
import { Button } from "./Button";
import { TotalAmount } from "./TotalAmount";
import { ExchangeRate } from "./ExchangeRate";

export const TotalContainer = styled("div")<{ invalid?: boolean }>`
  color: ${(props) => (props.invalid ? color.red : color.whiteTransparent)};
`;

export interface AccountProps {
  currency: string;
  amount: string;
  totalAmount: string;
  base?: boolean;
  exchangeRate?: ExchangeRateType;
  invalid?: boolean;
  onAmountChange: (text: string) => void;
  onAccountChange: (next?: boolean) => void;
  placeholder: string;
}

export const Account = React.memo(
  ({
    base,
    currency,
    amount,
    totalAmount,
    exchangeRate,
    invalid,
    onAmountChange,
    onAccountChange,
    placeholder,
  }: AccountProps) => {
    const inputRef = React.useRef<HTMLInputElement>(null);
    const handleClick = () => {
      const { current } = inputRef;
      if (current) {
        current.focus();
      }
    };

    const handleNext = () => {
      onAccountChange(true);
    };

    const handlePrev = () => {
      onAccountChange();
    };

    return (
      <div className="text-white bg-dark py-3" onClick={handleClick}>
        <div className="d-flex align-items-baseline">
          <Button className="btn btn-dark bg-dark" onClick={handlePrev}>
            〈
          </Button>
          <p>{currency}</p>
          <Button className="btn btn-dark bg-dark" onClick={handleNext}>
            〉
          </Button>
          <div className="d-flex flex-grow-1 justify-content-end w-25">
            {amount.length > 0 && <span>{base ? "-" : "+"}</span>}
            <AmountInput
              className="w-50 border-0 text-light bg-dark"
              placeholder={placeholder}
              value={amount}
              onChange={onAmountChange}
              inputRef={inputRef}
            />
          </div>
        </div>
        <div className="d-flex">
          <TotalContainer invalid={invalid}>
            <TotalAmount currency={currency} amount={totalAmount} />
          </TotalContainer>
          {exchangeRate && (
            <div className="d-flex flex-grow-1 justify-content-end">
              <ExchangeRate data={exchangeRate} rounded />
            </div>
          )}
        </div>
      </div>
    );
  }
);
