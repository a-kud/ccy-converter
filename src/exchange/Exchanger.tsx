import { Account } from "./Account";
import { AppContext } from "../AppContext";
import { Button } from "./Button";
import { ExchangeRate } from "./ExchangeRate";
import { getRate } from "../utils/api";
import { invertRate, round } from "../utils/number";
import React, { useMemo } from "react";

const REFRESH_INTERVAL = 1000000000;

export function Exchanger() {
  const { state, dispatch } = React.useContext(AppContext);
  const {
    accounts,
    baseAmount,
    quoteAmount,
    baseAccountIdx,
    quoteAccountIdx,
    rate,
  } = state;
  const baseAccount = accounts[baseAccountIdx];
  const quoteAccount = accounts[quoteAccountIdx];

  React.useEffect(() => {
    if (baseAccount.currency === quoteAccount.currency) {
      dispatch({ type: "RATE_CHANGED", payload: { rate: "1" } });
      return;
    }

    let timeout: number;
    const fetch = async () => {
      const newRate = await getRate(
        baseAccount.currency,
        quoteAccount.currency
      );
      dispatch({ type: "RATE_CHANGED", payload: { rate: newRate } });
      timeout = setTimeout(fetch, REFRESH_INTERVAL);
    };

    fetch();

    return () => clearTimeout(timeout);
  }, [baseAccount.currency, quoteAccount.currency, dispatch]);

  const handleBaseAmountChange = (text) => {
    dispatch({ type: "BASE_AMOUNT_CHANGED", payload: { amount: text } });
  };

  const handleQuoteAmountChange = (text) => {
    dispatch({ type: "QUOTE_AMOUNT_CHANGED", payload: { amount: text } });
  };

  const handleExchange = () => {
    dispatch({ type: "EXCHANGE_REQUEST", payload: {} });
  };

  const handleBaseAccountChange = (next?: boolean) => {
    dispatch({ type: "BASE_CHANGED", payload: { next } });
  };

  const handleQuoteAccountChange = (next?: boolean) => {
    dispatch({ type: "QUOTE_CHANGED", payload: { next } });
  };

  const exchangeRate = useMemo(
    () => ({
      base: quoteAccount.currency,
      quote: baseAccount.currency,
      rate: round(rate, 4),
    }),
    [quoteAccount, baseAccount, rate]
  );

  const exchangeRateInverted = useMemo(
    () => ({
      base: baseAccount.currency,
      quote: quoteAccount.currency,
      rate: invertRate(rate),
    }),
    [quoteAccount, baseAccount, rate]
  );

  const amountValid =
    parseFloat(baseAccount.amount) >= parseFloat(baseAmount ? baseAmount : "0");

  return (
    <div className="border border-3 border-dark">
      <div className="d-flex justify-content-between">
        <div className="d-flex flex-grow-1 justify-content-center align-items-center">
          <ExchangeRate data={exchangeRate} />
        </div>
        <Button
          onClick={
            baseAmount.length > 0 && amountValid ? handleExchange : undefined
          }
          className="btn btn-dark rounded-0"
        >
          Exchange
        </Button>
      </div>
      <Account
        base
        placeholder={"Amount to sell"}
        currency={baseAccount.currency}
        amount={baseAmount}
        totalAmount={baseAccount.amount}
        onAmountChange={handleBaseAmountChange}
        onAccountChange={handleBaseAccountChange}
        invalid={!amountValid}
      />
      <Account
        placeholder={"Amount to get"}
        currency={quoteAccount.currency}
        amount={quoteAmount}
        totalAmount={quoteAccount.amount}
        exchangeRate={exchangeRateInverted}
        onAmountChange={handleQuoteAmountChange}
        onAccountChange={handleQuoteAccountChange}
      />
    </div>
  );
}
