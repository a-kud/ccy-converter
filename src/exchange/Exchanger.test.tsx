import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import { AppState, initialState } from "../AppContext";
import { Account } from "./Account";
import { Exchanger } from "./Exchanger";

describe("Exchanger", () => {
  function mockAppContext(state: Partial<AppState> = {}) {
    const dispatch = jest.fn();
    jest.spyOn(React, "useContext").mockReturnValue({
      state: { ...initialState(), ...state, baseAmount: "100" },
      dispatch,
    });

    return { dispatch };
  }

  let wrapper;
  let dispatch;
  beforeEach(() => {
    jest.restoreAllMocks();
    dispatch = mockAppContext().dispatch;

    wrapper = shallow(<Exchanger />);
  });

  it("renders correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("dispatches BASE_AMOUNT_CHANGED", async () => {
    wrapper.find(Account).at(0).props().onAmountChange("foo");

    expect(dispatch).toBeCalledWith({
      type: "BASE_AMOUNT_CHANGED",
      payload: {
        amount: "foo",
      },
    });
  });

  it("dispatches EXCHANGE_REQUEST", async () => {
    wrapper.find("Button").at(0).props().onClick!();

    expect(dispatch).toBeCalledWith({
      type: "EXCHANGE_REQUEST",
      payload: {},
    });
  });
});
