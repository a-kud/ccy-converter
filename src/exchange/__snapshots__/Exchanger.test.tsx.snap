// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchanger renders correctly 1`] = `
<div
  className="border border-3 border-dark"
>
  <div
    className="d-flex justify-content-between"
  >
    <div
      className="d-flex flex-grow-1 justify-content-center align-items-center"
    >
      <ExchangeRate
        data={
          Object {
            "base": "USD",
            "quote": "EUR",
            "rate": "0",
          }
        }
      />
    </div>
    <Button
      className="btn btn-dark rounded-0"
      onClick={[Function]}
    >
      Exchange
    </Button>
  </div>
  <Memo()
    amount="100"
    base={true}
    currency="EUR"
    invalid={false}
    onAccountChange={[Function]}
    onAmountChange={[Function]}
    placeholder="Amount to sell"
    totalAmount="9999"
  />
  <Memo()
    amount=""
    currency="USD"
    exchangeRate={
      Object {
        "base": "EUR",
        "quote": "USD",
        "rate": "Infinity",
      }
    }
    onAccountChange={[Function]}
    onAmountChange={[Function]}
    placeholder="Amount to get"
    totalAmount="9000"
  />
</div>
`;
